plutus-core-interpreter/src/PlutusCoreInterpreter/CekMachine.hs

```diff
@@ -0,0 +1,118 @@
+computeCek env con (Wrap ann tyn ty term) = computeCek env (FrameWrap ann tyn ty : con) term
+computeCek env con (Unwrap _ term)        = computeCek env (FrameUnwrap : con) term
+computeCek env con tyAbs@TyAbs{}          = returnCek env con tyAbs
+computeCek env con lamAbs@LamAbs{}        = returnCek env con lamAbs
+computeCek env con constant@Constant{}    = returnCek env con constant
+computeCek _   _   Error{}                = EvaluationFailure
+computeCek env con var@(Var _ name)       = case lookupName name env of
+    Nothing           -> throw $ MachineException OpenTermEvaluatedMachineError var
+    Just (term, env') -> returnCek env' con term
+
+-- | The returning part of the CEK machine.
+returnCek :: Environment -> Context -> Plain Value -> EvaluationResult
+returnCek _   []                             res = EvaluationSuccess res
+returnCek env (FrameTyInstArg ty      : con) fun = instantiateEvaluate env con ty fun
+returnCek env (FrameApplyArg env' arg : con) fun = computeCek env' (FrameApplyFun env fun : con) arg
+returnCek env (FrameApplyFun env' fun : con) arg = applyEvaluate env' env con fun arg
+returnCek env (FrameWrap ann tyn ty   : con) val = returnCek env con $ Wrap ann tyn ty val
+returnCek env (FrameUnwrap            : con) dat = case dat of
+    Wrap _ _ _ term -> returnCek env con term
+    term            -> throw $ MachineException NonWrapUnwrappedMachineError term
+
+-- | Instantiate a term with a type and proceed.
+-- In case of 'TyAbs' just ignore the type. Otherwise check if the term is an
+-- iterated application of a 'BuiltinName' to a list of 'Value's and, if succesful,
+-- apply the term to the type via 'TyInst'.
+instantiateEvaluate
+    :: Environment -> Context -> Type TyName () -> Plain Term -> EvaluationResult
+instantiateEvaluate env con _  (TyAbs _ _ _ body) = computeCek env con body
+instantiateEvaluate env con ty fun
+    | isJust $ termAsPrimIterApp fun = returnCek env con $ TyInst () fun ty
+    | otherwise                      = throw $ MachineException NonPrimitiveInstantiationMachineError fun
+
+-- | Apply a function to an argument and proceed.
+-- If the function is a 'LamAbs', then extend the current environment with a new variable and proceed.
+-- If the function is not a 'LamAbs', then 'Apply' it to the argument and view this
+-- as an iterated application of a 'BuiltinName' to a list of 'Value's.
+-- If succesful, proceed with either this same term or with the result of the computation
+-- depending on whether 'BuiltinName' is saturated or not.
+applyEvaluate :: Environment -> Environment -> Context -> Plain Value -> Plain Value -> EvaluationResult
+applyEvaluate funEnv argEnv con (LamAbs _ name _ body) arg =
+    computeCek (extendEnvironment name arg argEnv funEnv) con body
+applyEvaluate funEnv _      con fun                    arg =
+    let term = Apply () fun arg in
+        case termAsPrimIterApp term of
```

Comment:
again, some judicious `ViewPatterns` might make this nicer.

Reply:
Could you show? It's maybe slightly verbose currently, but is kinda explicit about what is happening. Like "received a function, applied it to an argument, got a term, viewed it as a primitive iterated application". Easy to read.

Reply:
Ah, I take it back. You're using two variables from the original pattern, so it's not so simple.

This is the best I've got, not very compelling:
```
diff --git i/plutus-core-interpreter/src/PlutusCoreInterpreter/CekMachine.hs w/plutus-core-interpreter/src/PlutusCoreInterpreter/CekMachine.hs
index 6835aa5..4b96b94 100644
--- i/plutus-core-interpreter/src/PlutusCoreInterpreter/CekMachine.hs
+++ w/plutus-core-interpreter/src/PlutusCoreInterpreter/CekMachine.hs
@@ -9,6 +9,8 @@
 -- The CEK machine generates booleans along the way which might contain globally non-unique 'Unique's.
 -- This is not a problem as the CEK machines handles name capture by design.
 
+{-# LANGUAGE ViewPatterns #-}
+
 module PlutusCoreInterpreter.CekMachine
     ( EvaluationResult (..)
     , evaluateCek
@@ -96,17 +98,15 @@ instantiateEvaluate env con ty fun
 applyEvaluate :: Environment -> Environment -> Context -> Plain Value -> Plain Value -> EvaluationResult
 applyEvaluate funEnv argEnv con (LamAbs _ name _ body) arg =
     computeCek (extendEnvironment name arg argEnv funEnv) con body
-applyEvaluate funEnv _      con fun                    arg =
-    let term = Apply () fun arg in
-        case termAsPrimIterApp term of
-            Nothing                       ->
-                throw $ MachineException NonPrimitiveApplicationMachineError term
-            Just (IterApp headName spine) ->
-                case runQuote $ applyBuiltinName headName spine of
-                    ConstAppSuccess term' -> returnCek funEnv con term'
-                    ConstAppFailure       -> EvaluationFailure
-                    ConstAppStuck         -> returnCek funEnv con term
-                    ConstAppError err     -> throw $ MachineException (ConstAppMachineError err) term
+applyEvaluate funEnv _      con fun                    arg = let term = Apply () fun arg in
+    case term of
+        (termAsPrimIterApp -> Just (IterApp headName spine)) ->
+            case runQuote $ applyBuiltinName headName spine of
+                ConstAppSuccess term' -> returnCek funEnv con term'
+                ConstAppFailure       -> EvaluationFailure
+                ConstAppStuck         -> returnCek funEnv con term
+                ConstAppError err     -> throw $ MachineException (ConstAppMachineError err) term
+        _ -> throw $ MachineException NonPrimitiveApplicationMachineError term
 
 -- | Evaluate a term using the CEK machine. May throw a 'MachineException'.
 evaluateCek :: Term TyName Name () -> EvaluationResult
```

Reply:
Yeah, not too nice.